publish.go

```diff
@@ -66,29 +79,244 @@ func NewPublisherFromExisting(ctx context.Context,
 	headPublisher := head.NewPublisher()
 	startHeadPublisher(host, topic, headPublisher)
 
-	return &legPublisher{t, t.Close, host, headPublisher}, nil
+	publisher := &legPublisher{topic: t, onClose: t.Close, host: host, headPublisher: headPublisher}
+
+	adminSvr := publisher.startAdminServer(adminServerAddr)
+	publisher.adminSvr = adminSvr
+
+	return publisher, nil
 }
 
 func (lp *legPublisher) UpdateRoot(ctx context.Context, c cid.Cid) error {
-	log.Debugf("Published CID and addresses in pubsub channel: %s", c)
-	msg := message{
-		cid:   c,
-		addrs: lp.host.Addrs(),
-	}
-	err1 := lp.topic.Publish(ctx, encodeMessage(msg))
+	err1 := lp.publish(ctx, c)
 	err2 := lp.headPublisher.UpdateRoot(ctx, c)
 	if err1 != nil {
+		log.Errorw("Failed to publish root CID to pubsub channel", "root", c, "err", err1)
 		return err1
 	}
-	return err2
+
+	log.Infow("Published root CID to pubsub channel", c)
+	if err2 != nil {
+		log.Errorw("Failed to update root in head publisher", "root", c, "err", err2)
+		return err2
+	}
+	log.Errorw("Updated root in head publisher", "root", c)
+	return nil
+}
+
+func (lp *legPublisher) publish(ctx context.Context, c cid.Cid) error {
+	msg := message{
+		cid:   c,
+		addrs: lp.host.Addrs(),
+	}
+	return lp.topic.Publish(ctx, encodeMessage(msg))
 }
 
 func (lp *legPublisher) Close() error {
 	err1 := lp.headPublisher.Close()
+	if err1 != nil {
+		log.Errorw("Failed to close head publisher", "err", err1)
+	}
+
 	err2 := lp.onClose()
+	if err2 != nil {
+		log.Errorw("Failed to perform OnClose", "err", err2)
+	}
+
+	err3 := lp.adminSvr.Shutdown(context.Background())
+	if err3 != nil {
+		log.Errorw("Failed to shut down admin server", "err", err3)
+	}
 
 	if err1 != nil {
 		return err1
+	} else if err2 != nil {
+		return err2
 	}
-	return err2
+	return err3
+}
+
+func (lp *legPublisher) startAdminServer(addr string) *http.Server {
```

Comment:
This is not something that should be turned on by default as part of a legs publisher.

Maybe a http.mux for visibility could be provided in a utility package that an app can choose to link up, but the structure and choice of an http server really belongs in the application that uses the library for synchronization, not this library.

comparable wrappers like grpc / etc don't silently turn on an http admin servers when they're used.

Reply:
I do not intend to merge any of this; PR is in draft and created _purely_ for visibility.

Reply:
Added to PR description for clarity.

Reply:
Another reason this change is here and not in provider is because we can't upgrade lotus to the latest of index-provider until [this](https://github.com/filecoin-project/storetheindex/issues/130) is done.